p2p/ipld/validate.go

```diff
@@ -0,0 +1,99 @@
+package ipld
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	format "github.com/ipfs/go-ipld-format"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/lazyledger/nmt/namespace"
+
+	"github.com/lazyledger/lazyledger-core/types"
+)
+
+// ValidationTimeout specifies timeout for DA validation during which data have to be found on the network,
+// otherwise ErrValidationFailed is thrown.
+// TODO(@liamsi): define proper value
+const ValidationTimeout = time.Minute
+
+// ErrValidationFailed is returned whenever DA validation fails
+var ErrValidationFailed = errors.New("validation failed")
+
+// ValidateAvailability implements the protocol described in https://fc21.ifca.ai/papers/83.pdf.
+// Specifically all steps of the protocol described in section
+// _5.2 Random Sampling and Network Block Recovery_ are carried out.
+//
+// In more detail it will first create numSamples random unique coordinates.
+// Then, it will ask the network for the leaf data corresponding to these coordinates.
+// Additionally to the number of requests, the caller can pass in a callback,
+// which will be called on for each retrieved leaf with a verified Merkle proof.
+//
+// Among other use-cases, the callback can be useful to monitoring (progress), or,
+// to process the leaf data the moment it was validated.
+// The context can be used to provide a timeout.
+// TODO: Should there be a constant = lower bound for #samples
+func ValidateAvailability(
+	ctx context.Context,
+	api coreiface.CoreAPI,
+	dah *types.DataAvailabilityHeader,
+	numSamples int,
+	onLeafValidity func(namespace.PrefixedData8),
+) error {
+	// TODO(@Wondertan): Ensure data is fetched within one DAG session
+	ctx, cancel := context.WithTimeout(ctx, ValidationTimeout)
+	defer cancel()
+
+	squareWidth := uint32(len(dah.ColumnRoots))
+	samples := SampleSquare(squareWidth, numSamples)
+
+	type res struct {
+		data []byte
+		err  error
+	}
+	resCh := make(chan res, len(samples))
+	for _, s := range samples {
+		go func(s Sample) {
+			root, leaf, err := s.Leaf(dah)
+			if err != nil {
+				select {
+				case resCh <- res{err: err}:
+				case <-ctx.Done():
+				}
+				return
+			}
+
+			data, err := GetLeafData(ctx, root, leaf, squareWidth, api)
+			select {
+			case resCh <- res{data: data, err: err}:
+			case <-ctx.Done():
+			}
+		}(s)
```

Comment:
@adlerjohn, @liamsi. How many samples are expected to maximally validate a block? I am wondering if we need to limit the amount of spawned goroutines here.

Reply:
In a 64x64 ODS (128x128 EDS), you only need 64x64=4096 total shares to repair the entire block.

Reply:
forwarding the first paragraph of section 5.3 from the [Fraud and Data Availability Proofs](https://arxiv.org/pdf/1809.09044.pdf) paper for reference. (k = originalSquareWidth)

"In order for any share in the 2D Reed-Solomon matrix to be unrecoverable, then at least (k+ 1)2 out of (2k)2 shares must be unavailable (see Theorem 1). Thus When light clients receive a new block header from the network, they should randomly sample 0< s <(k+ 1)2 distinct shares from the extended matrix,and only accept the block if they receive all shares."